### PR TITLE
Add the feature of clicking on a request on a map

### LINKED ISSUE #85: Make map displayed requests clickable and show their details
make the requests on map clickable and connect them to accept or edit screen if user or owner

### PR DESCRIPTION
**Pull request by Funoma**
## ✨ Summary
Implement an interactive bottom sheet for request details with dynamic actions based on ownership.

**User Story**: #96
Addresses #85

---

### Core Implementation
* Add draggable, non-modal bottom sheet on map for request details
* Implement tabbed interface with 'Information' and 'Profile' tabs
* Map remains interactive when sheet is open
* Add dynamic action button based on request ownership:
  * Owner: Navigate to `EditRequestScreen`
  * Non-owner: Navigate to `AcceptRequestScreen`
* Display request information with styled components:
  * Title badge with primary color
  * Scrollable description
  * Start/End date cards
  * Status badge
  * Location with map pin icon

### Repository Enhancements
* Extended `RequestRepositoryFirestore` with ownership check:
  * `isCurrentUserOwner(requestId: String): Boolean`

### UI/UX Features
* **Draggable Bottom Sheet**:
  * Non-modal design allowing map interaction while open
  * Smooth height animation with customizable sizing
  * Drag handle for intuitive resizing
  * Close button for dismissal
* **Tabbed Navigation**:
  * Information tab: Full request details with dates, status, and location
  * Profile tab: Placeholder for requester profile (to be implemented with Firebase profile integration)
* **Responsive Design**:
  * All content scrollable together
  * Elements disappear gracefully when sheet is minimized

---

### Test Infrastructure
* Added comprehensive test tags like : DRAG_DOWN_MENU, REQUEST_TITLE,...

### UI Component Tests
* **MapsScreenTests** (`MapsTest.kt`)
  * Check if all components are displayed correctly, depending on current user.
---

## 📸 Video
[andStudMap.webm](https://github.com/user-attachments/assets/a3da5c1f-e9ab-4316-b48f-bb903a40304b)

## 📸 Screenshots
Capture d&#39;écran 2025-11-02 203451
Capture d&#39;écran 2025-11-02 203240

---

## 📝 Notes
* **Profile Tab**: Currently shows placeholder text. Full implementation pending Firebase profile integration
* **Sonar cloud issue**: There is a sonar cloud issue, but I think we can just accept it (I can't do it since I don't have permissions)
### AI Usage Disclosure
⚠️ **AI tools were used** to assist with code implementation and documentation writing for this PR.

Closes #85

### CODE REVIEW

File: app/src/main/java/com/android/sample/ui/map/MapViewModel.kt [hunk 1]
```diff
@@ -68,6 +70,30 @@ class MapViewModel(
     fetchAcceptedRequest()
   }
 
+  /** Refreshes the UI by updating the current request */
+  fun updateCurrentRequest(request: Request?) {
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/map/MapViewModel.kt]: > [!CAUTION]
> updateCurrentRequest() immediately calls isHisRequest() which launches a coroutine but doesn't wait for it. If the user > clicks another request before the first one finishes, you'll get stale data mixed with the new request.

File: app/src/main/java/com/android/sample/ui/map/MapViewModel.kt [hunk 2]
```diff
@@ -76,7 +102,23 @@ class MapViewModel(
     viewModelScope.launch {
       try {
         val requests = requestRepository.getAllRequests()
-        val loc = requests.firstOrNull()?.location ?: EPFL_LOCATION
+
+        val current = _uiState.value.currentRequest
+        // if the currentRequest has been deleted
+        if (current == null || requests.none { it.requestId == current.requestId }) {
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/map/MapViewModel.kt]: > [!NOTE]
> This may be an ineficcient search You do requests.none { it.requestId == current.requestId } (full scan), then 
> immediately do 
> requests.find { it.requestId == current.requestId } (another full scan). Do one search: requests.find { it.requestId == 
> current.requestId } and check if it's null.

File: app/src/main/java/com/android/sample/ui/map/ConstantMap.kt
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/map/ConstantMap.kt]: This is really good coding practice! you put all magic values in the same place.

File: app/src/androidTest/java/com/android/sample/request/RequestListTests.kt
Comment (Thibaud-Babin) [inline app/src/androidTest/java/com/android/sample/request/RequestListTests.kt]: lgtm

File: app/src/androidTest/java/com/android/sample/map/MapsTest.kt [hunk 3]
```diff
@@ -75,6 +141,93 @@ class MapsTest : BaseEmulatorTest() {
     mapsUtil.clickZoomOut()
     mapsUtil.clickZoomOut()
   }
+
+  // check that all components displayed correctly
+  @Test
+  fun clickOnOwnRequest() {
```
Comment (Thibaud-Babin) [inline app/src/androidTest/java/com/android/sample/map/MapsTest.kt]: Nice. Test names are clear and say what they test.

File: app/src/androidTest/java/com/android/sample/map/MapsTest.kt [hunk 4]
```diff
@@ -75,6 +141,93 @@ class MapsTest : BaseEmulatorTest() {
...
+    viewModel.updateCurrentRequest(request1)
+    composeTestRule.waitForIdle()
+
+    composeTestRule.onNodeWithTag(MapTestTags.DRAG).assertIsDisplayed()
```
Comment (Thibaud-Babin) [inline app/src/androidTest/java/com/android/sample/map/MapsTest.kt]: Using test tags instead of text matching is solid. Way more reliable.

File: app/src/androidTest/java/com/android/sample/map/MapsTest.kt [hunk 5]
```diff
@@ -75,6 +141,93 @@ class MapsTest : BaseEmulatorTest() {
...
+    composeTestRule
+        .onNodeWithTag(MapTestTags.BUTTON_DETAILS)
+        .performScrollTo()
+        .assertIsDisplayed()
+        .assertTextContains(ConstantMap.TEXT_EDIT)
+  }
+
+  // check if the name of the button has changed, and the button x works
+  @Test
+  fun clickOnAnotherRequestThenClose() {
```
Comment (Thibaud-Babin) [inline app/src/androidTest/java/com/android/sample/map/MapsTest.kt]: As with other functions in the test file, you're testing actual user flows (select request → check details → close). Not just isolated units. That is very good, nice !

File: app/src/androidTest/java/com/android/sample/map/MapsTest.kt [hunk 6]
```diff
@@ -25,13 +40,64 @@ class MapsTest : BaseEmulatorTest() {
...
+              Date(),
+              Date(System.currentTimeMillis() + 3_600_000),
+              listOf(),
+              emptyList(),
+              currentUserId)
+
+      repository.addRequest(request2)
+
+      signInUser(SECOND_USER_EMAIL, SECOND_USER_PASSWORD)
+      request1 =
```
Comment (Thibaud-Babin) [inline app/src/androidTest/java/com/android/sample/map/MapsTest.kt]: > [!TIP]
>That Request initialization is huge and repetitive. Could use a builder or factory function to make tests cleaner:
```kotlin
private fun createTestRequest(id: String, location: String, userId: String): Request {
    // Creates request with sensible defaults
}
```

File: app/src/androidTest/java/com/android/sample/map/MapsTest.kt
Comment (Thibaud-Babin) [inline app/src/androidTest/java/com/android/sample/map/MapsTest.kt]: You're tests are well written and easy to understand, very nice job and rigor!
I howevere have these notes:

-No negative tests : What happens if you click a deleted request? What if the request fails to load? No error cases    tested.
-constants/magic values
- test builder

**Review by Thibaud-Babin — CHANGES_REQUESTED** [review]
Good fundamentals overall solid null safety and defensive programming. What you implemented looks really nice and is well written, good job ! Also, yo managed to get no issue on Sonar cloud and you achieved more thean 90% nice coverage, outstanding! However, a few issues need fixing:

###  Critical
- **Race condition**: `updateCurrentRequest()` launches async ownership check without waiting. Stale data if users switch requests quickly.

###  Major
- **Broad exception catching**: Catch specific exceptions, not `Exception`. Also `e.message` can be null.
- **Ownership logic wrong**: When updating request, you keep old `isOwner` value. Shouldn't you recheck?
- **Inefficient search**: You call `requests.none()` then `requests.find()` on same list. Do one search.
- **Tight coupling**: `updateCurrentRequest()` always triggers ownership check. Should be separate.

###  Tests
- Test data creation is repetitive. Use factories instead.
- -**Magic values**: Strings, `3_600_000`, coordinates. Use named constants.
-Maybe some missing negative test cases (errors, deleted requests, edge cases).
- Magic numbers in test setup (`3_600_000`).

###  Good
- Clean null safety practices
- Good edge case thinking (deleted requests)
- Clear test naming
- Reasonable coverage

**Action:** Fix race condition and null assertions before merge. Extract magic values to constants.

**Comment by Funoma** [comment]
Fix the problems , and adapt the code in function of the review

**Review by Thibaud-Babin — APPROVED** [review]
You implemented and corrected the things I asked for. This is perfect thank you and nice job. Can't wait to see the code you will write next week 😼😼😼
